 # Assessing Data

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## 1. Quality
### 1.1. Visual Assessment

In [3]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

Các issues:
1. Table `patients`
    * Column `zip_cde` có format 5 digits nhưng một số entries có 4 digits.
    * Column `state` có hai formats full name và abbreviation.
    * Tim Neudord có height là 27 inches, là inconsistent với weight và BMI.
    * Typo (accuracy issue): Column `given_name` có typo Dsvid thay vì David.
2. Table `treatments`:
    * Column `hba1c_change` có hai NaN values.
    * Data format for dosage (validity issue): Hai columns `auralin` và `novodra` có format "digit, digit, u", và pandas sẽ coi đây là string thay vì number.
    * Missing data (completeness issue): Table `treatments` có 280 rows nhưng size of each treatment arm là 175 patients cho Auralin và Novodra arms, nên tổng records phải có là 174*2=350 thay vì 280.

3. Table `adverse_reactions`:
    * Data format for names (consistency issue): Hai columns `given_name` và `surname` đều là lowercase nhưng ở table `patients` thì bắt đầu bằng uppercase letters. Đây sẽ là issue nếu sau này ta join hai tables.

### 1.2. Progammatic Assessment

In [12]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

Another Quality Issue: Data Types
* Ở table `patients`, hai columns `assigned_sex` và `state` có type `object`, nhưng chúng nên có type `category` vì chúng chỉ có limit range of values. Tương tự, column `zip_code` là float nhưng nên là string và có type `object`, và column `birthdate` nên có type `datetime`
* Ở table `treatments`, hai columns `auralin` và `novodra` nên là integers.

Ngoài ra, `hba1c_change` đo effectiveness of insulin, có giá trị quanh $0.4$ đc coi là success, nên giá trị $0.9$ là implausible. Nhìn vào KQ từ `treatments.describe()`, 75 percentile cũng là $0.9$, và gap between 25 and 50 chỉ là $0.04$, trong khi giữa 50 and 75 là gần $0.6$. Nếu xem kĩ treatments table, ta sẽ thấy `hba1c_change=0.97` cho Elliot Richardson đã bị tính sai, nó nên là $0.47$ (accuracy issue).


In [13]:
patients.surname.value_counts()

surname
Doe            6
Jakobsen       3
Taylor         3
Ogochukwu      2
Tucker         2
              ..
Casárez        1
Mata           1
Pospíšil       1
Rukavina       1
Onyekaozulu    1
Name: count, Length: 466, dtype: int64

In [14]:
patients.address.value_counts()

address
123 Main Street             6
2778 North Avenue           2
2476 Fulton Street          2
648 Old Dear Lane           2
3094 Oral Lake Road         1
                           ..
1066 Goosetown Drive        1
4291 Patton Lane            1
4643 Reeves Street          1
174 Lost Creek Road         1
3652 Boone Crockett Lane    1
Name: count, Length: 483, dtype: int64

In [15]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

Ta thấy có 6 last names of "Doe" và 6 addresses of "123 Main Street." Khi dùng `.duplicated()` để kiểm tra kĩ ta thấy có nhiều John Doe's that live at 123 Main Street New York, New York, ZIP Code 12345 with the email johndoe@email.com. Đây là validity issue vì data ko tuân theo defined schema of one record per patient.

Tương tự, ta cũng thấy same address, "648 Old Deer Lane" associated with "Jakob Jakobson", and also "Jake Jakobson". Ta cũng thấy similar issues with "Pat Gersten " and "Sandy Taylor".

'Elizabeth Knudsen' cũng có ở duplicated column, nhưng đây ko là data quality issue vì 'Elizabeth Knudsen' có demographic information là NaN, khớp với other patients' records with missing address, city, state, etc. information.


In [16]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

Minimum value of weight in `patients` table là 48.8 pounds, khá thấp. Ta kiểm tra lại bằng cách tính BMI entries for this patient và thấy 48.8 thực tế là kilograms thay vì pounds. Đây là consistency issue with unit measurement of data.

In [18]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703  *weight_lbs / ( height_in*  height_in )
print(bmi_check)
patients[patients.surname == 'Zaitseva'].bmi

210    19.055827
dtype: float64


210    19.1
Name: bmi, dtype: float64

In [19]:
sum(treatments.auralin.isnull())

0

In [20]:
sum(treatments.novodra.isnull())

0

Ta thấy hai columns `auralin` và `novodra` ko có null entries, nhưng có các entries with dashes nên là null. Misrepresenting missing values là validity issue.

Tỏng kết quality issues:
1. `patients` table
    * zip code is a float not a string
    * zip code has four digits sometimes
    * Tim Neudorf height is 27 in instead of 72 in
    * full state names sometimes, abbreviations other times
    * Dsvid Gustafsson
    * Missing demographic information (address - contact columns)
    * Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
    * Multiple phone number formats
    * Default John Doe data
    * Multiple records for Jakobsen, Gersten, Taylor
    * kgs instead of lbs for Zaitseva weight
2. `treatments` table
    * missing HbA1c changes
    * the letter u in starting and ending doses for Auralin and Novodra
    * lowercase given names and surnames
    * missing records (280 instead of 350)
    * Erroneous datatypes (auralin and novodra columns)
    * Inaccurate HbA1c changes (4s mistaken as 9s)
    * Nulls represented as dashes (-) in auralin and novodra columns
3. `adverse_reactions` table
    * lowercase given names and surnames

## 2. Tidiness
### 2.1. Visual Assessment
Hai columns `auralin` và `novodra` vi phạm rule: mỗi variable tạo nên một column. Có ba variables: treatment (auralin or novodra), start dose (e.g. 41 units), and end dose (e.g. 48 units). Nên có ba columns nhưng hiện chỉ có hai, mỗi cái chứa hai variables.

### 2.2. Programmatic Assessment
Ta tìm duplicated column names:

In [21]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

Theo rule mỗi type of observational unit tạo nên một table, dataset này chỉ nên có hai tables: `patients` và `treatments`.
1. Table `patients` vẫn giữ nguyên như hiện tại.
2. Trong table `adverse_reaction`, column `adverse_reactions` nên đc đưa vào table `treatments`.